Spec Execution Engine behavior

protolambda opened the pull request:
This specs:
- The relation of deposits to the execution engine
- The Engine API, with the `transactions` field 
- Networking on engine side (different from rollup p2p PR, separate thing): no modification, but light configuration to be aware of
- Two types of sync the engine can operate with. Refer to rollup-node/driver for the block-by-block worst case. Spec happy-case abstractly (we re-use existing L1 behavior).

protolambda commented on exec-engine.md:
```diff
@@ -0,0 +1,146 @@
+# L2 Execution Engine
+
+The rollup mirrors the layer abstraction that the Merge introduces to Ethereum L1:
+- Consensus: *Determine* canonical chain, propagate *blocks*, maintain *system-state*
+- Execution: *Process* canonical chain, propagate *transactions*, maintain *user-state*
+
+The consensus task is implemented by the rollup-node, deriving most functionality from L1.
+The execution task is implemented by the engine, providing an equivalent user environment as L1, with deposits.
```
We can remove this from this doc, but I thought we should define this consensus-execution separation somewhere, as context for this document. Suggestions for intro-text welcome.

Also, regarding introduction: we should automate TOC generation for spec docs.

norswap reviewed (commented):
will write more comments later

protolambda commented on exec-engine.md:
```diff
@@ -0,0 +1,146 @@
+
+1. Engine API informs engine of chain head, unconditionally (part of regular node operation): 
+   - [`engine_executePayloadV1`][engine_executePayloadV1] is called with latest L2 block derived from L1.
+   - [`engine_forkchoiceUpdatedV1`][engine_forkchoiceUpdatedV1] is called with the current `unsafe`/`safe`/`finalized` L2 block hashes.
```
Both happen and are distinct calls, but it's up to the rollup-node spec to implement this in detail. We can specify sync here, but then we shouldn't duplicate it in the rollup-node spec.

norswap replied:
It's just that during my first read I thought these were things happened sequentially.

Maybe: "The engine API informs the the execution engine of a new chain head by sending one of these calls:"

I'm actually not sure what "unconditionally" nor what "part of regular node operation" mean here. (Regular node operation: as opposed to what?)

protolambda commented on exec-engine.md:
```diff
@@ -0,0 +1,146 @@
+
+The Engine interfaces abstract away transaction types with [EIP-2718][eip-2718].
+
+To support rollup-functionality, processing of a new Deposit `TransactionType` is implemented by the engine,
```
The former, we need some transaction type for deposits.

maurelian commented:
Metadata: fixes ENG-1729

protolambda commented on exec-engine.md:
```diff
@@ -0,0 +1,146 @@
+The rollup mirrors the layer abstraction that the Merge introduces to Ethereum L1:
+- Consensus: *Determine* canonical chain, propagate *blocks*, maintain *system-state*
+- Execution: *Process* canonical chain, propagate *transactions*, maintain *user-state*
+
```
I like the clear `A state <> B state` comparison, although on-chain is clear too. I'm removing this introduction section for now, we can put this type of thing in a different document.

protolambda commented on exec-engine.md:
```diff
@@ -0,0 +1,146 @@
+# L2 Execution Engine
+
+The rollup mirrors the layer abstraction that the Merge introduces to Ethereum L1:
```
Good idea, but will move it to future intro doc.

protolambda commented on exec-engine.md:
```diff
@@ -0,0 +1,146 @@
+- Consensus: *Determine* canonical chain, propagate *blocks*, maintain *system-state*
+- Execution: *Process* canonical chain, propagate *transactions*, maintain *user-state*
+
+The consensus task is implemented by the rollup-node, deriving most functionality from L1.
```
"Deriving" is confusing yes, but in L1/L2 design you can think of security properties that carry over or not. Deriving functionality = a L2 implementation that exposes the same thing, by using the L1 to guarantee the properties of the functionality (data availability is the main thing here).

I'm removing this section, an introduction type of doc is probably a better place for this, and then we can iterate on the exact wording in a different PR.

protolambda commented on exec-engine.md:
```diff
@@ -0,0 +1,146 @@
+- Part of sync towards a trusted block-hash (trusted through previous Engine API instruction)
+
+Deposits MUST never be consumed from the transaction pool.
+*The transaction pool can be disabled in a deposits-only rollup*
```
Yes, a sequencer would use the mempool to receive transactions (in addition to offering an RPC). If we distribute the sequencer work we can't rely on a centralized RPC to ingest transactions. Re-using the existing p2p transaction propagation seems like a natural fit.

norswap replied:
That does make sense, although having centralized transaction ingress enables some interesting MEV designs: https://github.com/ethereum-optimism/optimistic-specs/discussions/35 (i.e. sandwiched is impossible by design and backrunning can be administered by the sequencer). The open question is how well that generalizes in a multi-sequencer world, and that in turn depends on how exactly multi-sequencer coordination will happen. For instance if sequencer are going to proceed in a round-robin fashion, then centralized transaction submission to the current sequencer is still in the cards.

protolambda replied:
Transactions should be visible by all sequencers. MEV is not stoppable unless all sequencers decide not to reorder transactions. If you hide the transaction from other sequencers, by only submitting it to one sequencer through RPC, then you are still trusting the sequencer to not search for MEV.

protolambda commented on exec-engine.md:
```diff
@@ -0,0 +1,146 @@
+## Engine API
+
+*Note: the [Engine API][l1-api-spec] is in alpha, `v1.0.0-alpha.5`.
+There may be subtle tweaks, beta starts in a few weeks*
```
Haha, it's not intended as a heading though, just an italic sentence. I'll look through the lint problems after going through review comments.

protolambda commented on exec-engine.md:
```diff
@@ -0,0 +1,146 @@
+
+Within the rollup, the types of forkchoice updates translate as:
+- `headBlockHash`: block-hash of the head of the canonical chain. Labeled `"unsafe"` in user JSON-RPC.
+   Nodes may apply L2 blocks out of band ahead of time, and then reorg when L1 data conflicts.
```
> Reading this sequentially begs the quesiton: where are those L2 blocks coming from?

Not specified, yet. But when we move to a sequencer-enabled rollup then we can have p2p distribution of l2 blocks via rollup nodes. Without specifying that yet, I thought we'd at least need this kind of context to explain the difference between `headBlockHash` and `safeBlockHash.

The rollup-node always defines what's canonical, it never reads the engine. When there is no other input source than L1 (i.e. no p2p block propagation), then `headBlockHash == safeBlockHash`.

norswap replied:
In this case I would just do what I did with the rollup-node spec: say they're the same for now.

I **really, really** think we should avoid forward references, they make things much more confusing than they need to be. It's never clear if something is there for a reason or because "we'll need it later".

protolambda replied:
In the API they are not the same, they are two distinct fields. Making these backwards changes for the deposit rollup is not going to improve anything.

protolambda commented on exec-engine.md:
```diff
@@ -0,0 +1,146 @@
+- `safeBlockHash`: block-hash of the canonical chain, derived from L1 data, unlikely to reorg.
+- `finalizedBlockHash`: irreversible block-hash, matches lower boundary of the dispute-period.
+
+To support rollup-functionality, one backwards-compatible change is introduced
```
True, my bad, it's more common to combine words in Dutch. Whenever there's a dash I consider it to be one term. I'll update it to use less dashes.

protolambda commented on exec-engine.md:
```diff
@@ -0,0 +1,146 @@
+equivalent to the `transactions` field in [`ExecutionPayloadV1`][ExecutionPayloadV1]:
+> `Array of DATA` - Array of transaction objects, each object is a byte list (`DATA`) representing
+> `TransactionType || TransactionPayload` or `LegacyTransaction` as defined in [EIP-2718][eip-2718].
+
```
There's a chance it will be supported. And as long as it's compatible I don't feel like we have to add a new JSON-RPC method for this.

protolambda replied:
Also regarding JSON-RPC spec: for a very long time ethereum didn't have a good or complete JSON-RPC spec. The execution-apis repo *is* the most canonical spec (user type of API is incomplete, but they do host an interactive site that is generated based on spec docs). The engine API has no other better reference yet. The ethereum github wiki used to host some RPC docs, but those since moved to a separate wiki site (no longer maintained), and then moved to the execution-apis repo. `DATA` and `QUANTITY` are common between all those + lots of EIPs, I'm not sure if the format itself is specified anywhere though.

protolambda replied:
Execution-apis engine document lists it at the top under the "structures" header:
> Fields having DATA and QUANTITY types MUST be encoded according to the HEX value encoding section of Ethereum JSON-RPC API.
>
> Note: Byte order of encoded value having QUANTITY type is big-endian.

I'll add a link.

norswap replied:
Yes, the link referred to here is the one I used in the rollup node spec.

Did not know executions-apis was also the location for the regular JSON-RPC API, will update the link in the rollup node spec.

Regarding the call, I feel different things with different interface should have different names. I feel like the ETH specs is itself making a big point of that by having `V1` right there in the name. We should preserve that spirit. If it's changed to include the tx field, then we can revert to using the same name.

protolambda replied:
I introduced that very versioning into the Engine API spec (during merge retreat week), if it's compatible it does not need a major versioning change.

norswap replied:
So you're telling me that if, after the merge, the execution API spec were to change to add a field, the `V1` in the name would not be changed? If that's the case, then fine I suppose.

protolambda commented on exec-engine.md:
```diff
@@ -0,0 +1,146 @@
+
+No modifications to [`engine_executePayloadV1`][engine_executePayloadV1].
+Applies a L2 block to the engine state.
+
```
It doesn't build a payload, it just executes and validates an already complete payload.

norswap replied:
Ok! Note that this is the kind of details that would be really helpful to catch on PRs like the rollup node PR. Right now it does not mention that we need to call `engine_executePayloadV1` -- that's a big oversight!

protolambda replied:
We should iterate on the doc, and sync is definitely one of those things to adjust later, let's not try and write 100% of the doc in one go, the review there is already painfully long and it is good enough as-is.

protolambda commented on exec-engine.md:
```diff
@@ -0,0 +1,146 @@
+The execution-engine can acquire all data through the rollup-node, as derived from L1:
+*P2P networking is strictly optional.*
+
+However, to provide L1-equivalent experience, the P2P network functionality SHOULD be enabled, serving:
```
EVM-equivalence is like having the exact same EVM. L1-equivalance (or yellow paper equivalence etc.) is like having the same transaction-pool, discovery, state-sync, etc. There's more outside of EVM. Good point though, I'll update the wording.

protolambda commented on exec-engine.md:
```diff
@@ -0,0 +1,146 @@
+
+No modifications to L1 network functionality are required, except configuration:
+- `networkID`: Distinguishes the L2 network from L1 and testnets. 
+  Equal to the `chainID` of the rollup network.
```
will add links. Configuration is different per client implementation, but the IDs are based on EIP-155 and the relic in network identification on p2p from before eth-classic.

protolambda commented on exec-engine.md:
```diff
@@ -0,0 +1,146 @@
+No modifications to [`engine_getPayloadV1`][engine_getPayloadV1].
+Retrieves a payload by ID, prepared by `engine_forkchoiceUpdatedV1` when called with `payloadAttributes`.
+
+## Networking
```
This is a different networking stack than the rollup-node p2p stack. 0 changes required here. Keeping this high-level for now.

protolambda commented on exec-engine.md:
```diff
@@ -0,0 +1,146 @@
+  Equal to the `chainID` of the rollup network.
+- Activate Merge fork: Enables Engine API and disables propagation of blocks, 
+  as block-headers cannot be authenticated without consensus-layer.
+- Bootnode list: although DiscV5 is a shared network, bootstrap is faster through L2 nodes. 
```
Brevity is valuable too. Especially for things that are widely implemented and specified elsewhere. I'll try to find some links for the curious reader instead.

norswap replied:
I think we must keep in mind that the target audience includes people that are not hardcore into implementation of Ethereum.

Especially since we really encourage reusing existing components, we must assume users will not necessarily have read the specs for those. Even if they have, some context is valuable to get situated.

I'm personally confused (but have not read the networking specs at this point). I don't consider myself to be particularly brilliant, but I think it would be very dangerous to assume that our "target reader" will understand things better than I do at this stage.

I think generally we should err of the side of over-explaining rather than under-explaining. Or said otherwise, lower the expectations we have of our readers.

protolambda commented on exec-engine.md:
```diff
@@ -0,0 +1,146 @@
+## Sync
+
+The execution-engine can operate sync in different ways:
+- Happy-path: no special work for rollup-node, inform engine of the desired chain head, completes through engine P2P.
```
The point is that the engine does not determine consensus, it just follows whatever is desired. I'll clarify that it's the rollup node who says what is desired.

protolambda commented on exec-engine.md:
```diff
@@ -0,0 +1,146 @@
+
+The execution-engine can operate sync in different ways:
+- Happy-path: no special work for rollup-node, inform engine of the desired chain head, completes through engine P2P.
+- Worst-case: rollup-node detects stalled engine, completes sync purely from L1 data, no peers required.
```
The sub-section goes in a bit more detail, but ultimately this type of sync should be specified in the rollup-node doc.

protolambda commented on exec-engine.md:
```diff
@@ -0,0 +1,146 @@
+1. Engine API informs engine of chain head, unconditionally (part of regular node operation): 
+   - [`engine_executePayloadV1`][engine_executePayloadV1] is called with latest L2 block derived from L1.
+   - [`engine_forkchoiceUpdatedV1`][engine_forkchoiceUpdatedV1] is called with the current `unsafe`/`safe`/`finalized` L2 block hashes.
+2. Engine requests headers from peers, in reverse till the parent-hash matches the local chain
```
Client implementations may implement this slightly differently, it's just the general idea that counts. The L2 spec does not need any modifications here.

As noted in the spec:

> The exact P2P based sync is out of scope for the L2 specification:
the operation within the engine is the exact same as with L1 (although with an EVM that supports deposits).

protolambda commented:
Updated the lint rules to use `stern: true` instead of `strict: true` option for `MD013 - Line length`. I.e. only warn about line-length when there is white-space after the limit. Long URLs are not a problem.

And also added a `.gitignore` to not depend on global user gitignore for ignoring `node_modules` when installing/running the linter locally.

norswap commented:
> Updated the lint rules to use `stern: true` instead of `strict: true` option for `MD013 - Line length`. I.e. only warn about line-length when there is white-space after the limit. Long URLs are not a problem.
> 
> And also added a `.gitignore` to not depend on global user gitignore for ignoring `node_modules` when installing/running the linter locally.

This should really be in a separate PR, shouldn't it?

protolambda commented:
@norswap done #55 :+1:

norswap commented:
I'm thinking an intro in this particular file is still pretty useful to situate the user, but I can agree to add that in later, after this is merged.

norswap commented:
I'm okay with merging this. I think it should be made more precise and/or detailed on some points, but I also agree with proto that it's going to be best done in subsequent PRs.
